source/geometry/Quaternion.ooc

```diff
@@ -34,75 +34,57 @@ Quaternion: cover {
 	conjugate ::= This new(this real, -this imaginary)
 	inverse ::= this conjugate / (this real squared + this imaginary norm squared)
-	rotation ::= 2.0f * this logarithmImaginaryNorm
 	transform ::= this toFloatTransform3D()
 	// NOTE: Separation into parts assumes order of application X -> Y -> Z
-	rotationX: Float {
-		get {
-			result: Float
-			value := this w * this y - this z * this x
-			if ((value abs() - 0.5f) abs() < This precision)
-				result = 0.0f
-			else
-				result = (2.0f * (this w * this x + this y * this z)) atan2(1.0f - 2.0f * (this x squared + this y squared))
-			result
-		}
-	}
-	rotationY: Float {
-		get {
-			result: Float
-			value := this w * this y - this z * this x
-			if ((value abs() - 0.5f) abs() < This precision)
-				result = value sign * (Float pi / 2.0f)
-			else
-				result = ((2.0f * value) clamp(-1, 1)) asin()
-			result
-		}
-	}
-	rotationZ: Float {
-		get {
-			result: Float
-			value := this w * this y - this z * this x
-			if ((value abs() - 0.5f) abs() < This precision)
-				result = 2.0f * (this z atan2(this w))
-			else
-				result = (2.0f * (this w * this z + this x * this y)) atan2(1.0f - 2.0f * (this y squared + this z squared))
-			result
-		}
-	}
-	direction: FloatPoint3D {
-		get {
-			quaternionLogarithm := this logarithm
-			quaternionLogarithm imaginary / quaternionLogarithm imaginary norm
-		}
-	}
-	logarithm: This {
-		get {
-			result: This
-			norm := this norm
-			point3DNorm := this imaginary norm
-			if (point3DNorm != 0)
-				result = This new(norm log(), (this imaginary / point3DNorm) * ((this real / norm) acos()))
-			else
-				result = This new(norm, FloatPoint3D new())
-			result
-		}
-	}
-	exponential: This {
-		get {
-			result: This
-			point3DNorm := this imaginary norm
-			exponentialReal := this real exp()
-			if (point3DNorm != 0)
-				result = This new(exponentialReal * point3DNorm cos(), exponentialReal * (this imaginary / point3DNorm) * point3DNorm sin())
-			else
-				result = This new(exponentialReal, FloatPoint3D new())
-			result
-		}
-	}
+	rotationX: Float { get {
+		result: Float
+		value := this w * this y - this z * this x
+		if ((value abs() - 0.5f) abs() < This precision)
+			result = 0.0f
+		else
+			result = (2.0f * (this w * this x + this y * this z)) atan2(1.0f - 2.0f * (this x squared + this y squared))
+		result
+	}}
+	rotationY: Float { get {
+		result: Float
+		value := this w * this y - this z * this x
+		if ((value abs() - 0.5f) abs() < This precision)
+			result = value sign * (Float pi / 2.0f)
+		else
+			result = ((2.0f * value) clamp(-1, 1)) asin()
+		result
+	}}
+	rotationZ: Float { get {
+		result: Float
+		value := this w * this y - this z * this x
+		if ((value abs() - 0.5f) abs() < This precision)
+			result = 2.0f * (this z atan2(this w))
+		else
+			result = (2.0f * (this w * this z + this x * this y)) atan2(1.0f - 2.0f * (this y squared + this z squared))
+		result
+	}}
+	direction ::= (this logarithm imaginary) normalized
```

Comment:
Since this changed to a one-liner, it's probably better suited above line 40 with the others. After that, I'm happy.